db/hints/manager.cc

```diff
@@ -959,8 +959,11 @@ future<> manager::end_point_hints_manager::sender::wait_until_hints_are_replayed
         (**ptr).set_exception(abort_requested_exception());
     });
 
-    return (**ptr).get_future().finally([this, sub = std::move(sub)] {
-        manager_logger.debug("[{}] wait_until_hints_are_replayed_up_to(): returning afther the future was satisfied", end_point_key());
+    // When the future resolves, the endpoint manager is not guaranteed to exist anymore
+    // therefore we cannot capture `this`
+    auto ep = end_point_key();
+    return (**ptr).get_future().finally([sub = std::move(sub), ep] {
+        manager_logger.debug("[{}] wait_until_hints_are_replayed_up_to(): returning afther the future was satisfied", ep);
     });
```

Comment:
tip: use coroutines, and avoid manually managing promises and futures. Your life will be less exciting.

Reply:
In general, I try to use coroutines where I can, however here I think the `finally` method is nicer as it is shorter and simpler as there is no `finally` keyword in C++.

As for managing promises and futures... I couldn't find an abstraction which would allow me to have a priority queue with abortable promises which fully clean up after themselves in case of an abort, hence I wrote the `_replay_waiters`. Maybe it should be extracted to a new abstraction in order to make things clearer.